### PR TITLE
Support glm4 awq

### PR DESCRIPTION
**Pull request by AllentDan**
#1976

### CODE REVIEW

**Comment by akai-shuuichi** [comment]
Does this support quantization of glm4v？

**Comment by AllentDan** [comment]
> Does this support quantization of glm4v？

I only tested glm4-9b-chat model.

File: lmdeploy/turbomind/supported_models.py [hunk 1]
```diff
@@ -37,7 +37,8 @@
     MGMLlamaForCausalLM='llama',
     MiniGeminiLlamaForCausalLM='llama',
     # chatglm2/3, glm4
-    ChatGLMModel='glm4')
+    ChatGLMModel='glm4',
+    ChatGLMForConditionalGeneration='glm4')
```
Comment (lvhan028) [inline lmdeploy/turbomind/supported_models.py:41]: Can you share the huggingface repo id whose arch is defines as ChatGLMForConditionalGeneration?

File: lmdeploy/lite/quantization/awq.py [hunk 2]
```diff
@@ -57,6 +61,10 @@
     'DecoderLayer': {
         'self_attn.W_pack': ['self_attn.o_proj'],
         'mlp.up_proj': ['mlp.down_proj']
+    },
+    'GLMBlock': {
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/awq.py:65]: Can we remove the `GLMBlock` item?

File: lmdeploy/turbomind/supported_models.py (hunk 1 above)
Comment (lvhan028) [reply lmdeploy/turbomind/supported_models.py:41]: Please check if we need to update the unit tests

Comment (AllentDan) [reply lmdeploy/turbomind/supported_models.py:41]: It is https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/config.json#L9
After quantization (loading & saving) the model architecture is converted to `ChatGLMForConditionalGeneration`

File: lmdeploy/lite/quantization/awq.py (hunk 2 above)
Comment (AllentDan) [reply lmdeploy/lite/quantization/awq.py:65]: No, we can't do that. But we may remove the commented lines below it.

File: lmdeploy/lite/quantization/calibration.py [hunk 3]
```diff
@@ -192,24 +192,12 @@ def _forward(mod, *args, **kwargs):
                         self_attn.layer_idx = ori_idx
 
                         out = list(out)
-                        cache = out.pop(-1)
-
-                        key = cache.key_cache.pop(-1)
-                        value = cache.value_cache.pop(-1)
-
-                        k_obs.observe(key)
-                        v_obs.observe(value)
```
Comment (AllentDan) [inline lmdeploy/lite/quantization/calibration.py:201]: Here I removed offline kv cache quantization. cc @lvhan028

Comment (lvhan028) [reply lmdeploy/lite/quantization/calibration.py:201]: I see. Great job

File: lmdeploy/turbomind/supported_models.py (hunk 1 above)
Comment (lvhan028) [reply lmdeploy/turbomind/supported_models.py:41]: Why change it to `ChatGLMForConditionalGeneration`

Comment (AllentDan) [reply lmdeploy/turbomind/supported_models.py:41]: I did not change it. `AutoModel.from_pretrained` automatically forms it.

**Review by lvhan028 — APPROVED** [review]

**Comment by SquareMask** [comment]
你好，为什么我运行了还是RuntimeError: Currently, quantification and calibration of ChatGLMForConditionalGeneration are not supported. The supported model types are InternLMForCausalLM, InternLM2ForCausalLM, QWenLMHeadModel, Qwen2ForCausalLM, BaiChuanForCausalLM, BaichuanForCausalLM, LlamaForCausalLM, LlavaLlamaForCausalLM, MGMLlamaForCausalLM, InternLMXComposer2ForCausalLM. pip下的最新的0.5.1

**Comment by AllentDan** [comment]
这个PR合入后还没发版 @SquareMask 试试这个 https://github.com/zhyncs/lmdeploy-build/releases/tag/6367976

**Comment by SquareMask** [comment]
打扰问下还会支持W4A16吗 lmdeploy lite smooth_quant那个

**Comment by lvhan028** [comment]
lmdeploy  lite smooth_quant 是 w8a8 的。扩展w8a8支持的模型，会延后一些。等w8a8的推理性能提上来后，我们再补充

**Pull request merged**